pkg/trisa/keys/certs.go

```diff
@@ -0,0 +1,178 @@
+func FromX509KeyPair(cert *x509.Certificate, privateKey interface{}) (Key, error) {
+	return &Certificate{certs: cert, privateKey: privateKey}, nil
+}
+
+// FromProvider returns a Key from a deserialized trust provider. This is the most
+// common mechanism of loading a certificate from disk and is the most flexible.
+func FromProvider(certs *trust.Provider) (_ Key, err error) {
+	key := &Certificate{
+		privateKey: certs.GetKey(),
+	}
+
+	if key.certs, err = certs.GetLeafCertificate(); err != nil {
+		return nil, err
+	}
+	return key, nil
+}
+
+// Certificate is a wrapper for an x509 certificate (containing the public key) and
+// optionally for the private key associated with the certificate. This data structure
+// is used to create exchange keys using TRISA issued sealing-certificates. The public
+// key in the certificate will be used to send keys for sealing and the private key for
+// unsealing certificates.
+type Certificate struct {
+	certs      *x509.Certificate
+	privateKey interface{}
+}
+
+// Ensure the Certificate implements the Key interface.
+var _ Key = &Certificate{}
+
+// IsPrivate returns ture if there is a private key associated with the certificate.
+func (c *Certificate) IsPrivate() bool {
+	return c.privateKey != nil
+}
+
+// SealingKey is public key used to seal envelopes, usually an *rsa.PublicKey.
+func (c *Certificate) SealingKey() (interface{}, error) {
+	return c.certs.PublicKey, nil
+}
+
+// Proto returns the protocol buffer message with the public sealing key to exchange
+// with a remote counterparty to begin a TRISA transfer.
+func (c *Certificate) Proto() (msg *api.SigningKey, err error) {
+	msg = &api.SigningKey{
+		Version:            int64(c.certs.Version),
+		Signature:          c.certs.Signature,
+		SignatureAlgorithm: c.certs.SignatureAlgorithm.String(),
+		PublicKeyAlgorithm: c.PublicKeyAlgorithm(),
+		NotBefore:          c.certs.NotBefore.Format(time.RFC3339),
+		NotAfter:           c.certs.NotAfter.Format(time.RFC3339),
+	}
+
+	// TODO: should we marshal into PEM encoded form?
+	if msg.Data, err = x509.MarshalPKIXPublicKey(c.certs.PublicKey); err != nil {
+		return nil, fmt.Errorf("could not marshal pkix public key: %s", err)
+	}
+
+	return msg, nil
+}
+
+// UnsealingKey is the private key used to unseal envelopes, usually an *rsa.PrivateKey.
+// If no private key is available (IsPrivate() is false) then an error is returned.
+func (c *Certificate) UnsealingKey() (interface{}, error) {
+	if c.IsPrivate() {
+		return c.privateKey, nil
+	}
+	return nil, ErrNoPrivateKey
+}
+
+// PublicKeyAlgorithm refers to the public key algorithm of the x509.Certificate and
+// may be used to determine which cipher to use. Typically is "RSA".
+func (c *Certificate) PublicKeyAlgorithm() string {
+	return c.certs.PublicKeyAlgorithm.String()
+}
+
+// PublicKeySignature returns a unique identifier that can be used to manage public keys
+// and associate them with their counterpart private keys for unsealing.
```

Comment:
Suggest:

// PublicKeySignature returns a unique identifier that can be used to manage the public keys
// of one's counterparties; 
// or to manage one's own keys locally and associate them with the matching private keys for unsealing

Reply:
Does this help clarify?

```
// PublicKeySignature returns a unique identifier that can be used to manage public keys
// and associate them with their counterpart private keys for unsealing or to identify
// the keys in use by the counterparty for creating and sealing secure envelopes.
```